kernel/cooking.mli

```diff
@@ -23,7 +23,12 @@ open Constr
     (because the substitution are represented by their domain) but
     here, local definitions of the context have been dropped *)
 
-type abstr_inst_info
+type abstr_inst_info = {
```

Comment:
```suggestion
type abstr_inst_info = private {
```

Reply:
I really don't like exposing internal types that way. @LasseBlaauwbroek what do you want to achieve exactly? IIRC expmod_constr is just computing the abstraction of the term, so isn't it a variant of abstract_as_body?

Reply:
Reading your thread on Zulip, I believe you can get what you want by 1. calling  `abstract_as_body`, 2. computing the variable instance with `instance_of_cooking_cache`, 3. substituting the body. This is O(n) in the context size so not great if part of a critical piece of code, but if you only care about expressiveness and not efficiency, this should work, shouldn't it?

Reply:
@ppedrot Your proposal feels rather fragile to me, and seems to do a lot of duplicate work. This is partly because the API of `Cooking` doesn't contain a lot of documentation on what functions do exactly.
1. Use `abstract_as_body`
2. Remove all of the binders introduced by `abstract_as_body`. Is it guaranteed that `Array.length (instance_of_cooking_cache ..)` is equal to the number of binders introduced by `abstract_as_body`? (This does not seem to be documented somewhere.)
3. Convert de Bruijn indices back into named variables using `instance_of_cooking_cache`. I assume there is some substitution function for this somewhere?

It all seems much more inefficient and cumbersome than it was before with `expmod_constr`. I call these functions fairly often, so speed is not unimportant.

What is your objection to exposing the internal types? I'm happy to add the `private` modifier like @SkySkimmer proposes. Alternatively, we can also create functions of type
```
cooking_info -> Constant.t -> Id.t list * Univ.Instance.t
cooking_info -> MutInd.t -> Id.t list * Univ.Instance.t
```

Reply:
Exposing internal types allows stronger coupling with the calling code. This is annoying, especially given that the Cooking code has been rewritten several times in the past, and that would prevent future optimizations of the way we abstract constants. It's a pretty sensitive part, because this goes into vo files (it's bundled with opaque constants). Adding private is still exposing too much, because you know that the field are mere projections. At best, you should add these as opaque functions, as you suggest.

Regarding your point 2., Cooking is a quite internal API. Its only aim is to handle section abstraction, and I am not sure that your precise use case makes sense in general (e.g. assuming we tweak the internal representation at some point). If you can express it with the current API (which seems to be the case) I'd rather not expose anything else.

Reply:
> Cooking is a quite internal API.

I guess I don't really understand that. `Libobject.object_declaration` has a field `discharge_function` which seems to be pretty clearly part of the public API. And any plugin that maintains any kind of data containing a `EConstr.t` will need to somehow need to use the API in `Cooking` during section discharge. So then `Cooking` would need to properly support the operations that may be needed? Currently, it looks like the `Cooking` API is hyper-specialized for internal purposes, but that seems inappropriate to me.

I see three different solutions, which are all acceptable to me:
1. Reintroduce `expmod_constr`, allowing you to keep everything nice and private. Downside is that it would probably lead to some code duplication in `Cooking` (but otherwise that duplication would occur in my plugin, so that's also not great)
2. This PR, with the more opaque function signatures from my previous message if you like.
3. Your proposal, but then I'd request a more precise spec on the API. Currently, `instance_of_cooking_cache` just returns some array of `Constr.t`'s, without any explanation what this is or what it does... At the moment, it is not clear to me from the API how I should code up your solution...

> and I am not sure that your precise use case makes sense in general

I'm not sure I understand that. The idea of applying section variables and universes to constants in a term seems pretty well-defined to me, unless the whole concept of a section is going to change.

Reply:
> unless the whole concept of a section is going to change

I think that the long-term plan is to change the handling of section variables, which has been plaguing Coq for ages with a lot of subtle bugs (notably the perennial "clearing a section variable bug" and all its avatars). @herbelin tweaked the Cooking API with the idea in mind that at some point all constants in a section should be expanded already, and that we have some abstraction layer hiding this to the user. This is the converse of the current situation where we abstract constants at the end of sections.

Reply:
That seems like a sensible future direction to me. And if/when that happens, I'll be happy to adapt my plugin to such a new API (it will probably make my life much easier; these bugs that you mention are a serious pain that I've also ran into).

But in the mean time I'd still have some kind of intermediate solution because my plugin is currently broken. It seems to me that option (2) is by far the most conservative and easy approach. Is that something you can agree to?

Reply:
>The idea of applying section variables and universes to constants in a term seems pretty well-defined to me

Not really. In
~~~
Section S.
Variable x : nat.
Definition c := x.
~~~
you would produce `c x` which is not valid inside the section (`c` takes no arguments) and is not valid outside the section (`x` does not exist).

Reply:
In my situation it is valid outside of the section, because I'm keeping a context that includes section variables around. So to extend your example, I'm doing this:
```
Section S.
Variable x : nat.
Definition c := x.
x : nat, ... |- ... c ...                 <---- my data
```
After seciton closing, this gets transformed into
```
Definition c x := x.
x : nat, ... |- ... c x ...
```
This is a completely valid transformation as far as I can see.

Reply:
FWIW, here is a small snippet I cooked up (pun intended) using the current API that should mimick the old function.
```ocaml
let expmod_constr info c =
  let c = Cooking.abstract_as_body info c in
  let n_assums = List.length (Cooking.rel_context_of_cooking_cache info) in
  let ctx, c = Term.decompose_lam_n_assum n_assums c in
  let rec fold ctx args subst = match ctx, args with
  | [], [] -> subst
  | Context.Rel.Declaration.LocalAssum _ :: ctx, arg :: args ->
    fold ctx args (Esubst.subs_cons arg subst)
  | Context.Rel.Declaration.LocalDef (_, b, _) :: ctx, args ->
    let b = Vars.esubst Constr.lift subst b in
    fold ctx args (Esubst.subs_cons b subst)
  | _ :: _, [] | [], _ :: _ -> assert false
  in
  let args = Array.to_list @@ Cooking.instance_of_cooking_cache info in
  let subst = fold (List.rev ctx) args (Esubst.subs_id 0) in
  Vars.esubst Constr.lift subst c
```
As usual I probably screwed up the order of the arguments so this code might need a bit of tweaking, but that's the spirit of my proposal.

Reply:
Thanks @ppedrot, with some tweaks this indeed worked. As you note yourself though, this does need to go through the entire term for every variable in the section. That makes section closing quite slow sometimes for me. So from that perspective, it would still be good if this PR could be merged. But if you are adamantly against that, I guess this will do as well.

Reply:
Alternatively
~~~ocaml
let expmod_constr info c =
  let c = Cooking.abstract_as_type info c in
  let n_assums = Context.Rel.nhyps (Cooking.rel_context_of_cooking_cache info) in
  let _, c = Term.decompose_prod_n n_assums c in
  let args = Array.to_list @@ Cooking.instance_of_cooking_cache info in
  Vars.substl args c
~~~
because abstract_as_type will inline the localdefs

Reply:
Thanks @SkySkimmer, that is a nice simplication. (`args` needs to be reversed in order for it to be correct though)